### DISCUSSION TITLE
Provide `data` and `error` to Layout function

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [remix-run/remix#8846] [Proposals]
I'd like to suggest providing the new `Layout` function with (supported) access to `data` and `error`.

This would let us do the same kinds of things in `Layout` that we can already do today in `meta` and with our own custom `Document` components, such as:

```tsx
export function Layout({ data, error, children }: LayoutProps): ReactNode {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <title>
          {data != null
            ? data.title
            : isRouteErrorResponse(error)
              ? error.statusText
              : "Error"}
        </title>
        <Meta />
        <Links />
        <link
          rel="stylesheet"
          href={data != null ? data.stylesheet : defaultStylesheet}
        />
        <style
          dangerouslySetInnerHTML={{
            __html: `
              :root {
                --custom-property: ${data != null
                  ? data.customProperty
                  : defaultCustomProperty};
              }
            `,
          }}
        />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}
```

This shows a props-based API, but of course it could also be hook-based.

I think it's important because:

- `Layout` is probably the best place for global tags now, [for all the reasons already described in the docs](https://remix.run/docs/en/main/route/meta#global-meta).
- Global tags are often data-driven.
- Some tags, like `<style>` tags, can't be inserted using the `meta` API anyway.

For now the workaround is to not use `Layout`, I think. (Or you could try calling `useLoaderData()` and/or `useRouteError()` in your `Layout` component, but I bet that's going to break some things!)

**Comment by sergiodxa** [comment]
You can get the root data with `useRouteLoaderData<typeof loader>("root")` right now, the return type will be `SerializedFrom<typeof loader> | undefined` (it will be undefined in case of error.

I'm not sure if you can also call useRouteError too.

**Reply by aaronadamsCA** [reply]
Ah, `useRouteLoaderData` is a great idea for now, thanks for that!

It looks like you can call `useRouteError` in `Layout`, you get a console warning but it does work, and it's technically already type-safe since it returns `unknown`. I'm unsure if there are further impacts or risks.

If not, maybe removing that warning and adding some docs and tests is all there is to this, unless there's something I'm missing.

**Reply by brophdawg11** [reply]
What console warning are you referring to?

Right now, I think this is a fine pattern but I wouldn't be opposed to this idea of props if it turns out that the vast majority of folks are doing `data`/`error` forking in their Layout's:

```jsx
export function Layout({ children }: { children: React.ReactNode }) {
  let data = useRouteLoaderData("root");
  let error = useRouteError();
  ...
}
```

**Reply by aaronadamsCA** [reply]
> What console warning are you referring to?

@brophdawg11 I can't reproduce it, so who knows! 🤷

**Reply by brophdawg11** [reply]
For now, we added a note to the docs on this in https://github.com/remix-run/remix/pull/8958
